Docs: Flow control section (#754)
inngest/website: pages/docs/guides/debounce.mdx

> export const description = 'Avoid unnecessary function invocations by de-duplicating events over a sliding time window. Ideal for preventing wasted work when a function might be triggered in quick succession.'

Reviewer: ```suggestion
export const description = 'De-duplicate events over Avoid unnecessary function invocations by de-duplicating events over a sliding time window. Ideal for preventing wasted work when a function might be triggered in quick succession.'
```

# Debounce

> Debounce allows a function to delay execution until a series of events have stopped being received. This is useful for preventing wasted work when a function might be triggered in quick succession. Use cases for debounce include:

Reviewer: ```suggestion
Debounce delays function execution until a series of events are no longer received. This is useful for preventing wasted work when a function might be triggered in quick succession. Use cases for debounce include:
```

> * Prevent wasted work when handling events from user input that may change multiple times in a short time period

Reviewer: ```suggestion
* Preventing wasted work when handling events from user input that may change multiple times in a short time period.
```

> * Delay processing of noisy webhook events until they have stopped being received

Reviewer: ```suggestion
* Delaying processing of noisy webhook events until they are no longer received.
```

> * Ensuring that functions use the latest event within a series of updates (example: synchronization)

Reviewer: ```suggestion
* Ensuring that functions use the latest event within a series of updates (for example, synchronization).
```

## How to configure debounce

<CodeGroup forceTabs>
```ts
export default inngest.createFunction(
  {
    id: "handle-webhook",
    debounce: {
      key: "event.data.account_id",
      period: "5m",
    },
  },
  { event: "intercom/company.updated" },
  async ({ event, step }) => {

>     // This function will only be scheduled 5m after events have stopped being received with the same

Reviewer: ```suggestion
    // This function will only be scheduled 5 minutes after events are no longer received with the same
```

Author: Committing this, but it still feels weird - may be something to adjust later 🤔

    // `event.data.account_id` field.
    //
    // `event` will be the last event in the series received.
  }
);
```
</CodeGroup>

### Configuration reference

* `period` - The time delay to delay execution. The period begins when the first matching event is received.

> * `key` - An optional [expression](/docs/guides/writing-expressions) using event data to apply each limit too. Each unique value of the `key` has it's own limit, enabling you to rate limit function runs by any particular key, like a user id.

Reviewer: ```suggestion
* `key` - An optional [expression](/docs/guides/writing-expressions) using event data to apply each limit too. Each unique value of the `key` has its own limit, enabling you to rate limit function runs by any particular key, like a user ID.
```

## How it works

When a function is triggered, the debounce `period` begins. If another event is received that matches the function's trigger, the debounce `period` is reset.  This continues until no events are received for the debounce `period`.  Once the `period` has passed without any new events, the function is executed using the last event received.

<img
  src="/assets/docs/reference/debounce.png"
  width="800"
  height={337}
  quality="95"
  alt="Visualization of how debounce is applied"
/>

### Using a `key`

When a `key` is added, a separate debounce period is applied for each unique value of the `key` expression. For example if your `key` is set to `event.data.customer_id`, each customer would have their individual debounce period applied to functions run. Read [our guide to writing expressions](/docs/guides/writing-expressions) for more info.

## Comparison to rate limiting

> For use cases where you prefer to execute a function for the _first_ event received, consider using [rate limiting](/docs/guides/rate-limiting) instead. Rate limiting ensures that a function runs once for each `key` the *first* time an event is received, while debounce uses the *last* event during a specified period.

Reviewer: ```suggestion
If you prefer to execute a function for the _first_ event received, consider using [rate limiting](/docs/guides/rate-limiting) instead. Rate limiting ensures that a function runs once for each `key` the *first* time an event is received, while debounce uses the *last* event during a specified period.
```

## Combining with idempotency

> Debounce can be combined with [idempotency](/docs/guides/handling-idempotency#at-the-function-level-the-consumer) to ensure that once a function is debounced it does not run again.

Reviewer: ```suggestion
Debounce can be combined with [idempotency](/docs/guides/handling-idempotency#at-the-function-level-the-consumer) to ensure that once the debounced function has run, it does not run again.
```
